cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+        synchronized (lastNotificationLock) {
+          if (lastNotification == null ||
+            (lastNotification.getTimestamp()
+              + TimeUnit.SECONDS.toMillis(pollingDelay) <= System.currentTimeMillis())) {
+            long streamSize = pollStream();
+            lastNotification = new StreamSizeNotification(System.currentTimeMillis(), streamSize);
+          }
+        }
+      }
+      task.received(lastNotification);
+    }
+
+    /**
+     * Delete a scheduling task that is based on the data received by the stream referenced by {@code this} object.
+     */
+    public void deleteSchedule(Id.Program programId, SchedulableProgramType programType, String scheduleName) {
+      StreamSizeScheduleTask scheduleTask = scheduleTasks.remove(getScheduleId(programId, programType, scheduleName));
+      if (scheduleTask != null && scheduleTask.isRunning()) {
+        activeTasks.decrementAndGet();
+      }
+    }
+
+    /**
+     * Get the status a scheduling task that is based on the data received by the stream referenced by {@code this}
+     * object.
+     */
+    public ScheduleState scheduleTaskState(Id.Program programId, SchedulableProgramType programType,
+                                           String scheduleName) {
+      StreamSizeScheduleTask task = scheduleTasks.get(getScheduleId(programId, programType, scheduleName));
+      if (task == null) {
+        return ScheduleState.NOT_FOUND;
+      }
+      return task.isRunning() ? ScheduleState.SCHEDULED : ScheduleState.SUSPENDED;
+    }
+
+    /**
+     * @return true if this object does not reference any schedule, false otherwise
+     */
+    public boolean isEmpty() {
+      return scheduleTasks.isEmpty();
+    }
+
+    public Id.Stream getStreamId() {
+      return streamId;
+    }
+
+    @Override
+    public Type getNotificationFeedType() {
+      return StreamSizeNotification.class;
+    }
+
+    @Override
+    public void received(StreamSizeNotification notification, NotificationContext notificationContext) {
+      // We only pass the stream size notification to the schedule tasks if the notification
+      // came after the last seen notification
+      boolean send = false;
+      cancelPollingAndScheduleNext();
+      synchronized (lastNotificationLock) {
+        if (lastNotification == null || notification.getTimestamp() > lastNotification.getTimestamp()) {
+          send = true;
+          lastNotification = notification;
+        }
+      }
+      if (send) {
+        sendNotificationToActiveTasks(notification);
+      }
+    }
+
+    /**
+     * Send a {@link StreamSizeNotification} to all the active {@link StreamSizeSchedule} referenced
+     * by this object.
+     */
+    private void sendNotificationToActiveTasks(final StreamSizeNotification notification) {
+      List<StreamSizeScheduleTask> values;
+      synchronized (this) {
+        values = ImmutableList.copyOf(scheduleTasks.values());
```

Comment:
If the scheduleTasks are a concurrent map, then it is safe to iterate directly over the map. I think it is also better: Because if there is really a concurrent modification (say, a task is deleted) then you want to see that modification reflected here (otherwise you would call a deleted task).